ssstar/src/role_credentials_provider.rs

```diff
@@ -0,0 +1,161 @@
+//! This module provides credentials for role-based authorization
+//!
+//! The credentials are automatically refreshed after the specified duration that allows to use
+//! the this credential provider for long running jobs
+
+use crate::Result;
+use aws_credential_types::{
+    provider::{error::CredentialsError, ProvideCredentials},
+    Credentials,
+};
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+};
+use tokio::sync::Mutex;
+
+/// The configuration which is used to re-fresh credentials for the specified role-arn
+struct RefreshConfig {
+    /// Region which is used on `AssumeRole` operation
+    region: Option<String>,
+    /// Role-arn of the role which we expect to use to get temporary credentials
+    role_arn: String,
+    /// Session name of the role's credentials
+    role_session_name: String,
+    /// Duration of validity of temporary credentials which will be refreshed after the expire
+    role_session_duration: Option<i32>,
+}
+
+struct AutoRefreshingProviderInner {
+    /// Current valid credentials
+    credentials: Option<Credentials>,
+    /// Configuration for `AssumeRole` and generating temporary credentials
+    refresh_config: RefreshConfig,
+}
+
+impl AutoRefreshingProviderInner {
+    async fn refresh_credentials(&mut self) -> Result<()> {
```

Comment:
So every time when we need to make some AWS invocation this code will be executed? If yes, I think it could be more effective way to refresh credentials: e.g. it can be dome every X minutes but before existing credentials expire.

Reply:
I also thought it's called on every AWS call, but actually, it's called only when the credentials are expired, then aws-sdk calls it to retrieve the new credentials. I'll add comment about it